server/service/endpoint_utils.go

```diff
@@ -274,6 +295,13 @@ func makeDecoder(iface interface{}) kithttp.DecodeRequestFunc {
 			}
 		}
 
+		if isBodyDecoder {
```

Comment:
Why do we wait until here to decode the body with the `bodyDecoder` instead of on line 145 above? This may lead to some interesting behaviour with the order in which we handle different struct tags depending on whether the struct implements `bodyDecoder`. Basically, I'm proposing something like this

```
diff --git a/server/service/endpoint_utils.go b/server/service/endpoint_utils.go
index 949062544..30d5fbd6f 100644
--- a/server/service/endpoint_utils.go
+++ b/server/service/endpoint_utils.go
@@ -123,11 +123,6 @@ func makeDecoder(iface interface{}) kithttp.DecodeRequestFunc {
 		v := reflect.New(t)
 		nilBody := false
 
-		var isBodyDecoder bool
-		if _, ok := v.Interface().(bodyDecoder); ok {
-			isBodyDecoder = true
-		}
-
 		buf := bufio.NewReader(r.Body)
 		var body io.Reader = buf
 		if _, err := buf.Peek(1); err == io.EOF {
@@ -142,7 +137,11 @@ func makeDecoder(iface interface{}) kithttp.DecodeRequestFunc {
 				body = gzr
 			}
 
-			if !isBodyDecoder {
+			if bd, ok := v.Interface().(bodyDecoder); ok {
+				if err := bd.DecodeBody(ctx, body); err != nil {
+					return nil, err
+				}
+			} else {
 				req := v.Interface()
 				if err := json.NewDecoder(body).Decode(req); err != nil {
 					return nil, err
@@ -232,7 +231,7 @@ func makeDecoder(iface interface{}) kithttp.DecodeRequestFunc {
 			}
 
 			_, jsonExpected := f.Tag.Lookup("json")
-			if jsonExpected && nilBody && !isBodyDecoder {
+			if jsonExpected && nilBody {
 				return nil, errors.New("Expected JSON Body")
 			}
 
@@ -295,13 +294,6 @@ func makeDecoder(iface interface{}) kithttp.DecodeRequestFunc {
 			}
 		}
 
-		if isBodyDecoder {
-			bd := v.Interface().(bodyDecoder)
-			if err := bd.DecodeBody(ctx, body); err != nil {
-				return nil, err
-			}
-		}
-
 		return v.Interface(), nil
 	}
 }
```

Reply:
It's important that this comes last, after all non-body parameters have been decoded, so that the `DecodeBody` implementation has all this extra information, I mention this ordering in the comments. In this specific case, the body decoder for apply team specs must have access to the `force` flag.

Reply:
Ah, I did not see how the `Force` field was being used. Thanks for the clarification.